features/FEATURE_BLE/source/generic/GenericGap.tpp

```diff
@@ -1917,9 +1919,9 @@ ble_error_t GenericGap<PalGapImpl, PalSecurityManager, ConnectionEventMonitorEve
 
         _advertising_timeout.detach();
         if (maxDuration.value()) {
-            _advertising_timeout.attach_us(
+            _advertising_timeout.attach(
                 mbed::callback(this, &GenericGap::on_advertising_timeout),
-                durationCast<millisecond_t>(maxDuration).value()
+                std::chrono::microseconds(durationCast<millisecond_t>(maxDuration).value())
```

Comment:
Durations are a strange things in BLE, we have many different units which are not based on ms and its multiple adding an interface where you can convert to and from was natural. 

Replacing it with `chrono` may help us in solving the problem. We just have to think of Ann efficient way of solving the boundary issue. 

Thanks for spotting the issue, this is on the legacy compatibility path, we missed it.

Reply:
New suggestion - a simple non-conversion-operator method:

    auto Duration::valueChrono() const
    {
        return std::chrono::duration<Rep, std::ratio<TB, 1000000>::type>{duration};
    }

@pan- I saw your suggestion with the handling of "forever". Seems like overkill, with runtime cost? Can't really assume a Chrono user would treat `max()` as forever, and if there was special handling of some other "forever" value of the representation, they could check for it explicitly. The conversion maintains representation, so that's possible.

Reply:
I have [reverted](https://github.com/ARMmbed/mbed-os/compare/809a3d800b51c6c986bb39baff63832636987154..63b32905138634310117d3a392bf4c3e2c51e1ca) the section that require more complex changes for an elegant solution.

Reply:
Please review [this](https://github.com/ARMmbed/mbed-os/compare/63b32905138634310117d3a392bf4c3e2c51e1ca..2f9feaefffeb0a56e64ffdd1fd2212cb12393c46).

Reply:
Looks good at the use sites now, but I'd like @pan- to consider the `Duration` changes - naming (I'd have assumed this should be alongside the other `value` methods.), and justify need for the forever->max "conversion".

Reply:
@kjbracey-arm `Forever` is a special value used to not interrupt a process, sometimes it can be the highest value in the duration range but it may also be the lowest (for instance, 0).
I'd like the conversion to reflect this. Using the max value may not be the most appropriate thing to do, but it is better than converting something with an infinite duration to something with no duration at all. 

Maybe an assertion could be used to check that no conversion to chrono are made if the special value forever is being held. That value should be checked before in the code which is here to handle legacy devices behavior.

Reply:
My point is that the chrono types do not inherently have those semantics, again for any value. If you want those semantics, then you need to code them - I don't how coercing "forever value, whatever it may be" into "max" is better than preserving the forever value and assuming the same semantics. If you convert a type with "forever = 0" into chrono, then that value means forever if its value is 0.

I see that using "max" may make it not fail so badly when you forget to put the special "forever" handling in for the chrono section, but any conversion that doesn't put the special handling in is wrong. This is covering up a programming error, and with a runtime cost.

Yes, I think it would make sense to assert on a forever conversion. That should actually lead to no code, assuming it forces you to write

    if (myDuration.is_forever()) {
         myChrono = myChronoForever;
    } else {
         myChrono = myDuration.chronoValue();
    }

The assert inside the `chronoValue` shouldn't generate code, because it can see that the condition is already cleared by the if it's inside.  And if the forever representations match, even the if should vanish? Making it a direct assignment. Pretty certain it should.

Are you happy with the naming?

Reply:
I'am, I completely agree that less damaging doesn't mean its not wrong. The assertion seems to be the thing to do. If an `is_forever` member function is added it should be named `isForever` to keep the code consistent.

Reply:
Wasn't asking about the naming of my code snippet - that was just me guessing that you already had an `is_forever` or something. Surely you do. And `chronoValue` was my misremembering of my original `valueChrono` suggestion.

I was repeating my earlier question about the name `getChronoDuration` as it currently stands. Doesn't seem to fit to me.

Reply:
`valueChrono` is a good replacement for `getChronoDuration`.

Reply:
@kjbracey-arm @pan- 
Could you please review [this](https://github.com/ARMmbed/mbed-os/compare/2f9feaefffeb0a56e64ffdd1fd2212cb12393c46..705752e1609b039eeb4fe9576cc6fe20d97502e1)?
Thanks.